### PR TITLE
[SPARSE] Add support for rocSPARSE backend

### PR DESCRIPTION
**Pull request by Rbiessy**
# Description

Add support for the rocSPARSE backend.

Depends on https://github.com/oneapi-src/oneMKL/pull/527 and https://github.com/oneapi-src/oneMKL/pull/532.

Rendered docs: [docs.zip](https://github.com/user-attachments/files/16363101/docs.zip)

# Checklist

## All Submissions

- [x] Do all unit tests pass locally? [amd_log.txt](https://github.com/user-attachments/files/16363103/amd_log.txt) [nvidia_log.txt](https://github.com/user-attachments/files/16363105/nvidia_log.txt) [intel_log.txt](https://github.com/user-attachments/files/16363106/intel_log.txt)
  - Tested the rocSPARSE backend on MI210 with oneAPI 2024.2 compiler + [Codeplay AMD plugin](https://developer.codeplay.com/products/oneapi/amd/home/) + ROCm 6.1.0

- [x] Have you formatted the code using clang-format?

### CODE REVIEW

**Comment by Rbiessy** [comment]
Tests log on W6800: [amd_w6800_log.txt](https://github.com/user-attachments/files/16430518/amd_w6800_log.txt)

**Comment by gajanan-choudhary** [comment]
@Rbiessy, now that #527 is merged in, could you please rebase this branch against `main`, resolve conflicts, and force-push when you can so that I can start reviewing this PR?

**Comment by Rbiessy** [comment]
Yes, this is in progress! I expect I will need a few days at least. I'm aiming to merge the PR by the end of the year.

**Comment by Rbiessy** [comment]
@gajanan-choudhary I have updated the PR with recent changes from cuSPARSE. Note that it moved almost all the content from `cusparse_task.hpp` to `common_launch_task.hpp` with almost no changes.
New test logs: 
[log_mi210.txt](https://github.com/user-attachments/files/17601859/log_mi210.txt)
[log_a100.txt](https://github.com/user-attachments/files/17601862/log_a100.txt)
[log_pvc.txt](https://github.com/user-attachments/files/17601863/log_pvc.txt)

File: README.md
Comment (gajanan-choudhary) [inline README.md:68]: Please also add rocSPARSE version info in the table later in this README file in the table in the [Product and Version Information](https://github.com/oneapi-src/oneMKL?tab=readme-ov-file#product-and-version-information) section

File: docs/domains/sparse_linear_algebra.rst
Comment (gajanan-choudhary) [inline docs/domains/sparse_linear_algebra.rst]: Wow, this is quite severe, but seems to be a legitimate issue on `rocSPARSE` side right now.

Comment (gajanan-choudhary) [reply docs/domains/sparse_linear_algebra.rst]: If and when they fix this issue, though, will it be easy for us to make changes (with a version check of course) that correctly performs the operations rather than throwing an `unimplemented` exception?

File: src/sparse_blas/backends/cusparse/cusparse_handles.cpp
Comment (gajanan-choudhary) [inline src/sparse_blas/backends/cusparse/cusparse_handles.cpp:25]: Why is a cuSPARSE change coming in here?

Comment (gajanan-choudhary) [reply src/sparse_blas/backends/cusparse/cusparse_handles.cpp:25]: Is it related to the `common_launch_task.hpp` to keep common code between cuSPARSE and rocSPARSE backends?

File: src/sparse_blas/backends/rocsparse/rocsparse_task.hpp
Comment (gajanan-choudhary) [inline src/sparse_blas/backends/rocsparse/rocsparse_task.hpp]: Other than cusparse -> rocsparse, this file, rocsparse/rocsparse_task.hpp appears to be nearly identical to cusparse/cusparse_task.hpp. Is there anything more than that in this file?

File: docs/domains/sparse_linear_algebra.rst
Comment (Rbiessy) [reply docs/domains/sparse_linear_algebra.rst]: Yes, it's easy to fix on oneMath side. The issue is also referenced in this comment: https://github.com/oneapi-src/oneMKL/pull/544/files#diff-3b8c1c2c71abd54f8f90f43415c2f17b2a7fdb81c2b882c210f3cba56b4679adR63
One would just need to remove the `used` member, its 2 usages below as well as the `mark_used` method.

File: src/sparse_blas/backends/cusparse/cusparse_handles.cpp
Comment (Rbiessy) [reply src/sparse_blas/backends/cusparse/cusparse_handles.cpp:25]: It's just as you said, `cusparse_task.hpp` used to include `cusparse_scope_handle.hpp` but it's not needed anymore (with the already merged changes to use the ScopeHandle object less often) and not possible anymore with the introduction of the more generic `common_launch_task.hpp`.

File: src/sparse_blas/backends/rocsparse/rocsparse_task.hpp
Comment (Rbiessy) [reply src/sparse_blas/backends/rocsparse/rocsparse_task.hpp]: Yes, they are almost identical. I'm not sure I understand, why do you ask?

File: README.md
Comment (Rbiessy) [reply README.md:68]: Thanks for catching that, done in https://github.com/oneapi-src/oneMKL/pull/544/commits/dc22051919dff1cf439fa9a0962ce1ab17e8efab

File: src/sparse_blas/backends/rocsparse/rocsparse_handles.hpp
Comment (gajanan-choudhary) [inline src/sparse_blas/backends/rocsparse/rocsparse_handles.hpp:91]: In the documentation, `docs/domains/sparse_linear_algebra.rst`, you've written:
> - The CSR format requires the column indices to be sorted within each row.

So shouldn't we be handling both CSR and COO formats here instead of just COO?

File: src/sparse_blas/backends/rocsparse/rocsparse_task.hpp
Comment (gajanan-choudhary) [reply src/sparse_blas/backends/rocsparse/rocsparse_task.hpp]: I was just confirming for my understanding to make sure I wasn't missing anything.

File: src/sparse_blas/backends/rocsparse/rocsparse_scope_handle.cpp
Comment (gajanan-choudhary) [inline src/sparse_blas/backends/rocsparse/rocsparse_scope_handle.cpp:129]: Is it unchanged between the removed PI versus new UR APIs? Isn't there a `sycl::detail::ur::contextSetExtendedDeleter` somewhere that you need to place under `#ifdef ONEAPI_ONEMKL_PI_INTERFACE_REMOVED #else #endif...` ?

File: src/sparse_blas/backends/rocsparse/rocsparse_handles.cpp
Comment (gajanan-choudhary) [inline src/sparse_blas/backends/rocsparse/rocsparse_handles.cpp:78]: I wonder if we should throw an exception if `dvhandle->size < size` instead of `!=`. E.g., users may want to pad the`sycl::buffer` for some reason with zeros in the end, which wouldn't work with the current `if` condition.

Comment (gajanan-choudhary) [reply src/sparse_blas/backends/rocsparse/rocsparse_handles.cpp:78]: Nevermind, this is `set_dense_vector_data` used for replacing/switching out arrays in an existing handle. The condition is fine then.

File: src/sparse_blas/backends/rocsparse/rocsparse_handles.cpp
Comment (gajanan-choudhary) [inline src/sparse_blas/backends/rocsparse/rocsparse_handles.cpp:39]: Can we add a check here for `sycl::buffer` case that checks `buffer->size() >= size` and throws an exception otherwise?

File: src/sparse_blas/backends/rocsparse/rocsparse_handles.cpp
Comment (gajanan-choudhary) [reply src/sparse_blas/backends/rocsparse/rocsparse_handles.cpp:78]: Although maybe it doesn't hurt to add the check both here and in `init_xxx_data`

File: src/sparse_blas/backends/rocsparse/operations/rocsparse_spmv.cpp
Comment (gajanan-choudhary) [inline src/sparse_blas/backends/rocsparse/operations/rocsparse_spmv.cpp]: Is there a reason why this variable specifically is kept and captured by reference _outside_ `compute_functor` while other members of `spmv_descr` are created inside the functor? If it is because `buffer_size` is used later in an `if` condition, I'd prefer it if we change this approach and capture everything by `[=]` in `compute_functor`, even though it would mean replacing the last `buffer_size > 0` with `spmv_descr->temp_buffer_size > 0`. I know that what you have right now is expected to work in this particular case (because of the condition that `spmv_descr` must live as long as `spmv` is being called), it is normally a bad idea to capture variables by reference in SYCL functors that are going to be running asynchronously unless you have an immediate `event.wait()` (e.g., like what we are doing in `spmv_buffer_size()` function).

**Review by gajanan-choudhary — APPROVED** [review]
LGTM other than some minor change requests. This PR was a lot easier to review having reviewed #527. Thanks for the fantastic work!

File: src/sparse_blas/backends/rocsparse/rocsparse_scope_handle.cpp
Comment (Rbiessy) [reply src/sparse_blas/backends/rocsparse/rocsparse_scope_handle.cpp:129]: The `contextSetExtendedDeleter` is still in the `pi` namespace see https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/detail/ur.hpp#L100. As the comment says there are plans to deprecate it. The oneAPI core team at Codeplay started to change the scopeContextHandle mechanism for BLAS in https://github.com/oneapi-src/oneMKL/pull/609. They will make sure that oneMKL Interface will keep working if/when `contextSetExtendedDeleter` is deprecated.
If we ever need to change the scopeContextHandle mechanism, I would rather that we do it for both cuSPARSE and rocSPARSE backends in a separate PR.

File: src/sparse_blas/backends/rocsparse/rocsparse_handles.cpp
Comment (Rbiessy) [reply src/sparse_blas/backends/rocsparse/rocsparse_handles.cpp:78]: The check to verify that the buffer size is compatible with the handle size is done elsewhere, see https://github.com/oneapi-src/oneMKL/blob/develop/src/sparse_blas/generic_container.hpp#L164
This is a common place that the other backends also use. That check will run for `init_*_data` functions.

File: src/sparse_blas/backends/rocsparse/rocsparse_handles.cpp
Comment (Rbiessy) [reply src/sparse_blas/backends/rocsparse/rocsparse_handles.cpp:39]: This is already done, see my comment above: https://github.com/oneapi-src/oneMKL/pull/544#discussion_r1863590450

File: src/sparse_blas/backends/rocsparse/rocsparse_handles.hpp
Comment (Rbiessy) [reply src/sparse_blas/backends/rocsparse/rocsparse_handles.hpp:91]: Good point, I added this check in https://github.com/oneapi-src/oneMKL/pull/544/commits/b6518c90d986b7d65f37aed368fcaeac18adad79
I've also made sure we also have tests on AMD and Nvidia which are skipped when a property is not set. The runtime example also need to set this property now.

File: src/sparse_blas/backends/rocsparse/operations/rocsparse_spmv.cpp
Comment (Rbiessy) [reply src/sparse_blas/backends/rocsparse/operations/rocsparse_spmv.cpp]: Sure that's fine, done in https://github.com/oneapi-src/oneMKL/pull/544/commits/c53ef8d2cc49f9ec695b1d01bd49d238513ac42d

**Comment by Rbiessy** [comment]
Thanks for the review Gajanan!

For completeness, logs on MI210: [log_mi210.txt](https://github.com/user-attachments/files/17981584/log_mi210.txt)
and A100: [log_a100.txt](https://github.com/user-attachments/files/17981589/log_a100.txt)

FYI we will get a second approval and I aim to merge this after the oneMKL Interface renaming PR.

**Comment by Rbiessy** [comment]
Our AMD HW is unavailable for a while unfortunately. I will work on fixing the conflicts as soon as I can test the changes.

**Comment by Rbiessy** [comment]
I fixed conflicts with the oneMath renaming. New test log on MI210: [onemath_mi210.txt](https://github.com/user-attachments/files/18026827/onemath_mi210.txt)

**Review by s-Nick — APPROVED** [review]

**Pull request merged**
